### PR TITLE
Implement virtual keyword

### PR DESCRIPTION
**Pull request by Marenz**
refs #5424

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `b12074d467808c516988fd18b1280718f0b70024`:
```
Coding style error:
 libsolidity/analysis/ContractLevelChecker.cpp:257: if (function->overrides())

```
Please check that your changes are working as intended.

**Comment by Marenz** [comment]
Note: this is currently possible:
```
contract C {
	function() external virtual fp;
	function() external override fp2;
	function() external override virtual fp3;
}
```

**Comment by Marenz** [comment]
feedback from the chat:
* dissallow virtual for public variables
* emit warning for virtual in interfaces not only when overriding
* in Token.h, wrong `=>` change

**Comment by Marenz** [comment]
Updated:
* hopefully fixed all tests
* added change log
* fixed token.h problem

not fixed: any public variable specific code. I want to fix that behavior in an extra PR

File: test/libsolidity/syntaxTests/viewPureChecker/interface.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/viewPureChecker/interface.sol]: This we don't want to test here, do we?

File: test/libsolidity/syntaxTests/nameAndTypeResolution/425_interface_implement_public_contract.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/425_interface_implement_public_contract.sol]: Same as below - there should be one test for virtual in interfaces and this warning should be avoided anywhere else.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/424_using_interface_complex.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/424_using_interface_complex.sol]: Remove it.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/359_illegal_override_payable_nonpayable.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/359_illegal_override_payable_nonpayable.sol]: Here we should *add* virtual, since the test is about a different reason for the overload to be wrong.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/358_illegal_override_payable.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/358_illegal_override_payable.sol]: Also add ``virtual`` here.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/059_illegal_override_visibility.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/059_illegal_override_visibility.sol]: Add ``virtual`` here instead.

File: test/libsolidity/syntaxTests/inheritance/override/remove_view.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/inheritance/override/remove_view.sol]: Also add ``virtual`` here.

File: test/libsolidity/syntaxTests/inheritance/override/override_interface_multiple.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/inheritance/override/override_interface_multiple.sol]: Remove ``virtual``

File: test/libsolidity/syntaxTests/inheritance/override/add_view.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/inheritance/override/add_view.sol]: Add ``virtual`` here.

File: test/libsolidity/SolidityParser.cpp
Comment (ekpyron) [inline test/libsolidity/SolidityParser.cpp]: Curious - this should no longer be reserved, should it? Unrelated to this PR, but we might want to check that!

File: libsolidity/parsing/Parser.h
Comment (ekpyron) [inline libsolidity/parsing/Parser.h]: Haha, I just wanted to suggest to ditch the ``is`` part, but we, of course, can't just name it ``bool virtual;`` :-D. So fine :-). You might want to member-initialize to ``false``, though.

File: libsolidity/parsing/Parser.cpp
Comment (ekpyron) [inline libsolidity/parsing/Parser.cpp]: Yeah, ok, this is WIP I guess - (and to be changed to not allowing virtual for any variable declarations).

File: libsolidity/ast/AST.h
Comment (ekpyron) [inline libsolidity/ast/AST.h]: Member initialize to ``false`` (yeah, it's redundant, but still better to do it and we said we'd want to always do it for PODs).

File: libsolidity/ast/AST.h
Comment (ekpyron) [inline libsolidity/ast/AST.h]: Haha, I *would* suggest to (redundantly) member-initialize this to ``false`` as well, but looking at the lines above this would be somewhat superficial :-D. Can you just member-initialize all of these ``bool``s to ``false`` anyways and also ``m_location`` below to ``Location::Unspecified``? We have to start doing it consistently some time, so why not now...

Comment (ekpyron) [reply libsolidity/ast/AST.h]: I haven't done this consistently myself in the recent past, though, I must admit, since I just read the not-member-initialized ``FunctionDefinition::kind`` above :-). But still.

File: libsolidity/analysis/ContractLevelChecker.cpp
Comment (ekpyron) [inline libsolidity/analysis/ContractLevelChecker.cpp]: We could even think about adding a ``isVirtualOrInterfaceFunction`` that checks both - that might make it less likely to miss the fact that interface functions are implicitly virtual despite ``isVirtual`` returning ``false``? But yeah: as you like, I'd be fine without that.

File: liblangutil/Token.h
Comment (ekpyron) [inline liblangutil/Token.h]: If you want you can move ``Abstract`` out of the reserved keywords as well here - otherwise we should do it separately - apparently that was missed in the abstract PR. Or is this one just not rebased on top of that one?

**Review by ekpyron — CHANGES_REQUESTED** [review]
This looks good - only some minor suggestions apart from the tests - for the tests we should take care that they - as far as possible - check what they are supposed to and *only* check what they are supposed to.

File: libsolidity/parsing/Parser.cpp
Comment (ekpyron) [reply libsolidity/parsing/Parser.cpp]: You could still consider to just remove these changes from this PR instead of adding this here only to remove it in a follow-up PR... but as you like!

File: liblangutil/Token.h
Comment (Marenz) [reply liblangutil/Token.h]: I believe it was forgotten. I'll move it.

File: libsolidity/analysis/ContractLevelChecker.cpp
Comment (Marenz) [reply libsolidity/analysis/ContractLevelChecker.cpp]: Would you then get rid of `isVirtual`?

File: Changelog.md
Comment (chriseth) [inline Changelog.md]: I think this still fits the first line

File: libsolidity/analysis/ContractLevelChecker.cpp
Comment (chriseth) [inline libsolidity/analysis/ContractLevelChecker.cpp]: `{` not needed.

Comment (chriseth) [reply libsolidity/analysis/ContractLevelChecker.cpp]: Actually not even nested if needed.

File: libsolidity/analysis/ContractLevelChecker.cpp
Comment (chriseth) [reply libsolidity/analysis/ContractLevelChecker.cpp]: `hasVirtualSemantics` - `isMarkedVirtual`

Comment (chriseth) [reply libsolidity/analysis/ContractLevelChecker.cpp]: Or better in line with coding stile: `virtualSemantics` - `markedVirtual`

File: libsolidity/analysis/ContractLevelChecker.cpp
Comment (chriseth) [inline libsolidity/analysis/ContractLevelChecker.cpp]: Is the error shown on `_super` or on `_function`? Should be shown on `_super`.

File: libsolidity/analysis/ContractLevelChecker.cpp
Comment (chriseth) [inline libsolidity/analysis/ContractLevelChecker.cpp:789]: Does this generate duplicate errors now?

File: libsolidity/parsing/Parser.cpp
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]:
```suggestion
	while (true)
```

File: libsolidity/analysis/ContractLevelChecker.cpp
Comment (ekpyron) [reply libsolidity/analysis/ContractLevelChecker.cpp]: Or maybe this will be the new ``canBeOverridden`` :-)? But yeah, I think it'd be best to split into two, but I don't mind the exact naming - e.g. ``virtualSemantics`` and ``markedVirtual`` would work for me.

File: libsolidity/analysis/ContractLevelChecker.cpp
Comment (Marenz) [reply libsolidity/analysis/ContractLevelChecker.cpp:789]: well, it can print similar errors for overrides, but the secondary error location would then show to a different overloaded function.

Example:

```
abstract contract C
{
	function foo() internal;
}
abstract contract A
{
	function foo() internal;
}
contract D is C, A {
	function foo() internal override(A, C) {}
}

Error: Trying to override non-virtual function. Did you forget to add "virtual"?
 --> test.sol:10:2:
  |
10 |    function foo() internal override(A, C) {}
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Note: Overridden function is here:
 --> test.sol:3:2:
  |
3 |     function foo() internal;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^

Error: Trying to override non-virtual function. Did you forget to add "virtual"?
 --> test.sol:10:2:
  |
10 |    function foo() internal override(A, C) {}
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Note: Overridden function is here:
 --> test.sol:7:2:
  |
7 |     function foo() internal;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^


```

File: libsolidity/analysis/ContractLevelChecker.cpp
Comment (ekpyron) [reply libsolidity/analysis/ContractLevelChecker.cpp]: Or ``overridable()`` for that matter - but in the end that's the semantics of being ``virtual``, isn't it?

**Comment by chriseth** [comment]
Please two tests where C inherits from contracts A, B, and both A and B define a function that is virtual in A and non-virtual in B and

1. the function is defined in C
2. the function is not defined in C

File: libsolidity/parsing/Parser.cpp
Comment (ekpyron) [reply libsolidity/parsing/Parser.cpp]: This should be added to ``check_style.sh`` I guess...

**Comment by Marenz** [comment]
uuuuu[..]uuupdated

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `e834982b117ef34cd10e46584bb9487f65755259`:
```
Coding style error:
 libsolidity/analysis/StaticAnalyzer.cpp:232: while(expr)

```
Please check that your changes are working as intended.

**Comment by Marenz** [comment]
moved rule addition to own PR: https://github.com/ethereum/solidity/pull/7683

**Comment by Marenz** [comment]
rebased & pushed to get the CI fixes

**Comment by Marenz** [comment]
Hmm seems we're still missing one test fix

**Comment by ekpyron** [comment]
@Marenz could you rebase again? That should now fix the tests, shouldn't it?

**Comment by Marenz** [comment]
Updated changelog
updated breaking changes descr.
moved warning for virtual in interfaces to type checker

File: libsolidity/parsing/Parser.cpp
Comment (ekpyron) [inline libsolidity/parsing/Parser.cpp]: This line isn't really necessary anymore now that we have it member-initialized... Do you still want to keep it? For this line I'd be fine with keeping it...
Actually the line below is worse and what caught my eye here :-), because it suggests ``result.overrides`` is a plain pointer, which it isn't... it so happens that ``shared_ptr`` has a constructor for ``nullptr_t``, so it works, but if anything, I'd rather use ``result.overrides = {};`` or ``result.overrides.reset();``... or actually remove the line...

So yeah, I'd suggest to remove this line and the one below altogether... what do you think?

**Review by ekpyron — COMMENTED** [review]
This looks good to me now! One minor comment and I haven't yet checked, whether we have sufficient test coverage now. Also I haven't looked at the docs in detail yet.

File: libsolidity/parsing/Parser.cpp
Comment (Marenz) [reply libsolidity/parsing/Parser.cpp]: I am fine with removing both lines.

Comment (Marenz) [reply libsolidity/parsing/Parser.cpp]: updated

**Comment by Marenz** [comment]
All change requests have been made, this is till waiting for the final review

File: libsolidity/analysis/ContractLevelChecker.cpp
Comment (chriseth) [reply libsolidity/analysis/ContractLevelChecker.cpp:789]: Maybe it would be better to swap primary and secondary locations?

Comment (chriseth) [reply libsolidity/analysis/ContractLevelChecker.cpp:789]: Ok, maybe the order is correct, but we could move the `Did you forget to add "virtual"` to the secondary location? As it is now, it could suggest that you should add `virtual` to the overriding function.

Comment (Marenz) [reply libsolidity/analysis/ContractLevelChecker.cpp:789]: That has already been done here: https://github.com/ethereum/solidity/pull/7647/files#diff-b1ba2bbc8a967e12c1721caf45edb970R283

The example is from an earlier time

Comment (Marenz) [reply libsolidity/analysis/ContractLevelChecker.cpp:789]: For clarification, this is how the output with the current code looks:

```
Error: Trying to override non-virtual function. Did you forget to add "virtual"?
 --> test.sol:4:2:
  |
4 |     function foo() internal;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^
Note: Overriding function is here:
  --> test.sol:11:2:
   |
11 |    function foo() internal override(A, C) {}
   |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Error: Trying to override non-virtual function. Did you forget to add "virtual"?
 --> test.sol:8:2:
  |
8 |     function foo() internal;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^
Note: Overriding function is here:
  --> test.sol:11:2:
   |
11 |    function foo() internal override(A, C) {}
   |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


```

**Review by chriseth — APPROVED** [review]

**Pull request merged**
